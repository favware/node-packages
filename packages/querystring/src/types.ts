/**
 * Options for the Stringify function
 * @typedef {StringifyOptions} StringifyOptions tringify Options
 * @property {string} [separator] separator the querystring should get
 * @property {string} [equals] equals sign the querystring should use
 * @property {boolean} [includeQuestion] whether the querystring should be prefixed with a `?`
 */
export type StringifyOptions = {
  separator?: string;
  equals?: string;
  includeQuestion?: boolean;
};

/**
 * Options for the Parse function
 * @typedef {ParseOptions} ParseOptions Parse Options
 * @property {string} [separator] separator the querystring has
 * @property {string} [equals] equals sign the querystring has
 */
export type ParseOptions = {
  separator?: string;
  equals?: string;
};

/**
 * Generic object for type definitions
 * @typedef {QuerystringObject} QuerystringObject Generic object
 * @property {any} key String based key with value of any type
 * @private
 */
export type QuerystringObject = {
  [key: string]: string | number | boolean | any; // eslint-disable-line @typescript-eslint/no-explicit-any
};

/**
 * Extension to generic object defining an absolute error property
 * @typedef {QuerystringError} QuerystringError Generic Error Object
 * @property {string} err Error with a string value
 * @private
 */
export type QuerystringError = {
  err: string;
} & QuerystringObject;

/**
 * Predefined potential sources for errors
 * @typedef {FunctionSource} FunctionSource Source Reference Enum
 * @property {FunctionSource.Parse} Parse for when the source is {@link parse}
 * @property {FunctionSource.Stringify} Stringify for when the source is {@link stringify}
 * @private
 */
export enum FunctionSource {
  Parse = 'parse',
  Stringify = 'stringify',
}