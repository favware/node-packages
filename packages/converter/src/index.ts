export * from './converter';
export * from './helpers';
