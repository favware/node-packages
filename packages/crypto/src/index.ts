import crypto from './crypto';

export * from './util';
export * from './crypto';
export * from './StreamBinary';
export default crypto;
