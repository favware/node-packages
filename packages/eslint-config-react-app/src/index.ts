/**
 * Standard ESLint config for Favware projects using React + TypeScript
 * @example
 * ```json
 * {
 *   "extends": "@favware/eslint-config-react-app"
 * }
 * ```
 */
export const config = {
  root: true,
  extends: ['@favware', 'plugin:import/typescript'],
  plugins: ['import', 'jsx-a11y', 'react', 'react-hooks'],
  parserOptions: {
    project: './tsconfig.eslint.json',
    sourceType: 'module',
    ecmaVersion: 2019,
    ecmaFeatures: { jsx: true },
    jsx: true
  },
  settings: {
    react: {
      version: 'detect'
    }
  },
  rules: {
    // Overwriting config from @favware/eslint-config to better suite React applications
    'no-invalid-this': 'off',
    '@typescript-eslint/explicit-module-boundary-types': 'off',
    '@typescript-eslint/no-explicit-any': 'off',
    '@typescript-eslint/naming-convention': [
      'error',
      {
        selector: 'default',
        format: ['camelCase', 'PascalCase', 'UPPER_CASE'],
        leadingUnderscore: 'forbid',
        trailingUnderscore: 'forbid',
        filter: {
          regex: '^(?:VariableDeclarator|AssignmentExpression)$',
          match: false
        }
      },
      {
        selector: 'parameter',
        format: ['camelCase', 'PascalCase', 'UPPER_CASE'],
        leadingUnderscore: 'allow'
      },
      {
        selector: 'property',
        modifiers: ['readonly'],
        format: ['PascalCase']
      },
      {
        selector: 'enumMember',
        format: ['PascalCase']
      },
      {
        selector: 'typeLike',
        format: ['PascalCase']
      }
    ],

    // Implement desired rules that also get set by eslint-config-react-app
    'no-extra-label': 'warn',
    'no-iterator': 'warn',
    'no-label-var': 'warn',
    'no-labels': ['warn', { allowLoop: true, allowSwitch: true }],
    'no-lone-blocks': 'warn',
    'no-loop-func': 'warn',
    'no-mixed-operators': [
      'warn',
      {
        groups: [
          ['&', '|', '^', '~', '<<', '>>', '>>>'],
          ['==', '!=', '===', '!==', '>', '>=', '<', '<='],
          ['&&', '||'],
          ['in', 'instanceof']
        ],
        allowSamePrecedence: false
      }
    ],
    'no-multi-str': 'warn',
    'rest-spread-spacing': ['warn', 'never'],
    'no-restricted-properties': [
      'error',
      {
        object: 'require',
        property: 'ensure',
        message:
          'Please use import() instead. More info: https://facebook.github.io/create-react-app/docs/code-splitting'
      },
      {
        object: 'System',
        property: 'import',
        message:
          'Please use import() instead. More info: https://facebook.github.io/create-react-app/docs/code-splitting'
      }
    ],
    'getter-return': 'warn',

    // Configure eslint-plugin-import
    'import/first': 'error',
    'import/no-amd': 'error',
    'import/no-webpack-loader-syntax': 'error',

    // Configure eslint-plugin-react
    'react/forbid-foreign-prop-types': [
      'warn',
      {
        allowInPropTypes: true
      }
    ],
    'react/jsx-no-comment-textnodes': 'warn',
    'react/jsx-no-duplicate-props': 'warn',
    'react/jsx-no-target-blank': 'warn',
    'react/jsx-no-undef': 'error',
    'react/jsx-pascal-case': [
      'warn',
      {
        allowAllCaps: true,
        ignore: []
      }
    ],
    'react/jsx-uses-react': 'warn',
    'react/jsx-uses-vars': 'warn',
    'react/no-children-prop': 'warn',
    'react/no-access-state-in-setstate': 'warn',
    'react/no-danger': 'error',
    'react/no-danger-with-children': 'warn',
    'react/no-direct-mutation-state': 'warn',
    'react/no-is-mounted': 'warn',
    'react/no-this-in-sfc': 'warn',
    'react/no-typos': 'error',
    'react/react-in-jsx-scope': 'error',
    'react/require-render-return': 'error',
    'react/style-prop-object': 'warn',
    'react/void-dom-elements-no-children': 'warn',

    // Configure eslint-plugin-jsx-a11y
    'jsx-a11y/accessible-emoji': 'warn',
    'jsx-a11y/alt-text': 'warn',
    'jsx-a11y/anchor-has-content': 'warn',
    'jsx-a11y/anchor-is-valid': [
      'warn',
      {
        aspects: ['noHref', 'invalidHref']
      }
    ],
    'jsx-a11y/aria-activedescendant-has-tabindex': 'warn',
    'jsx-a11y/aria-props': 'warn',
    'jsx-a11y/aria-proptypes': 'warn',
    'jsx-a11y/aria-role': [
      'warn',
      {
        ignoreNonDOM: true
      }
    ],
    'jsx-a11y/aria-unsupported-elements': 'warn',
    'jsx-a11y/heading-has-content': 'warn',
    'jsx-a11y/iframe-has-title': 'warn',
    'jsx-a11y/img-redundant-alt': 'warn',
    'jsx-a11y/no-access-key': 'warn',
    'jsx-a11y/no-distracting-elements': 'warn',
    'jsx-a11y/no-redundant-roles': 'warn',
    'jsx-a11y/role-has-required-aria-props': 'warn',
    'jsx-a11y/role-supports-aria-props': 'warn',
    'jsx-a11y/scope': 'warn',

    // Configure eslint-plugin-react-hooks
    'react-hooks/exhaustive-deps': 'warn',
    'react-hooks/rules-of-hooks': 'error'
  }
};

module.exports = config;
export default config;
