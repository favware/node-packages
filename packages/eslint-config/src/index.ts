/**
 * Default ESLint configuration for Favware projects
 * @example
 * ```json
 * {
 *   "extends": "@favware"
 * }
 * ```
 */
export const config = {
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: './tsconfig.eslint.json',
    jsx: true,
    sourceType: 'module',
    ecmaVersion: 2019,
    ecmaFeatures: { jsx: false }
  },
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended'
  ],
  env: {
    node: true,
    es6: true,
    jest: true,
    browser: true,
    commonjs: true
  },
  rules: {
    // TypeScript
    '@typescript-eslint/no-non-null-assertion': 'off',
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/consistent-type-definitions': ['error', 'interface'],
    '@typescript-eslint/interface-name-prefix': ['error', { prefixWithI: 'never' }],
    '@typescript-eslint/no-for-in-array': 'warn',
    '@typescript-eslint/no-unnecessary-qualifier': 'warn',
    '@typescript-eslint/no-unnecessary-type-assertion': 'warn',
    '@typescript-eslint/no-useless-constructor': 'warn',
    '@typescript-eslint/prefer-for-of': 'warn',
    '@typescript-eslint/prefer-includes': 'warn',
    '@typescript-eslint/unified-signatures': 'warn',

    // Best Practices
    eqeqeq: ['error', 'smart'],
    'no-else-return': 'warn',
    'no-eval': 'error',
    'no-floating-decimal': 'error',
    'no-implicit-coercion': 'error',
    'no-invalid-this': 'error',
    'no-iterator': 'error',
    'no-lone-blocks': 'error',
    'no-loop-func': 'error',
    'no-return-assign': 'error',
    'no-sequences': 'error',
    yoda: 'error',

    // ECMAScript 6
    'no-duplicate-imports': 'error',
    'prefer-arrow-callback': 'error',
    'prefer-template': 'error'
  }
};

module.exports = config;
export default config;
