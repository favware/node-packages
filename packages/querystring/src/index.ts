/** Options for the {@link stringify} function */
export interface StringifyOptions {
  /** Separator the querystring should get */
  separator?: string;
  /** Equals sign the querystring should use */
  equals?: string;
  /** Whether the querystring should be automatically prefixed with a `?` */
  includeQuestion?: boolean;
  /**
   * Whether to use [`encodeURIComponent`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent)
   * on the parameters
   *
   * @default true
   */
  encodeUriComponents?: boolean;
}

/** Options for the {@link parse} function */
export interface ParseOptions {
  /** Separator the querystring has */
  separator?: string;
  /** Equals sign the querystring has */
  equals?: string;
}

/**
 * Checks what kind of primitive type the property is and transforms accordingly
 *
 * @remarks
 * Used for stringify
 *
 * @param v Input to check for primitive type
 */
export function stringifyPrimitive<T extends unknown>(v: T, encode: boolean | undefined): string {
  let primitive: string;

  switch (typeof v) {
    case 'string':
      primitive = v;
      break;
    case 'boolean':
      primitive = v ? 'true' : 'false';
      break;
    case 'number':
      primitive = isFinite(v) ? v.toString() : '';
      break;
    default:
      primitive = '';
  }

  return encode ? encodeURIComponent(primitive) : primitive;
}

/**
 * Type safely checks if an object includes a given property without using the prototype
 * @param obj Object to analyze
 * @param prop Property to check for
 */
export function objectHasProperty<O extends object>(obj: O, prop: keyof O) {
  return obj && prop in obj;
}

/**
 * Stringifies an object
 *
 * @param obj Object to stringify
 * @param options Options for the stringify, see {@link StringifyOptions}
 * @returns The stringified query object
 * @example
 * ```ts
 * stringify({prop: 'value', prop2: 'value2'})
 * ```
 *
 * @example
 * ```ts
 * stringify({prop: 'value', prop2: 'value2'}, {separator: '&', equals: '=', includeQuestion: false})
 * ```
 *
 * @example
 * ```ts
 * stringify({prop: 'value', prop2: 'value2'}, {separator: '&&', equals: '=', includeQuestion: true})
 * ```
 */
export function stringify<I>(
  obj: I,
  options: StringifyOptions = { separator: '&', equals: '=', includeQuestion: false, encodeUriComponents: true }
): string {
  try {
    if (!obj || Object.keys(obj).length <= 0) throw new Error('object_is_empty');
    if (typeof obj !== 'object') throw new Error('input_not_object');
    if (!objectHasProperty(options, 'separator')) options.separator = '&';
    if (!objectHasProperty(options, 'equals')) options.equals = '=';
    if (!objectHasProperty(options, 'includeQuestion')) options.includeQuestion = false;
    if (!objectHasProperty(options, 'encodeUriComponents')) options.encodeUriComponents = true;

    const combinedValues: string[] = [];
    for (const [key, value] of Object.entries(obj)) {
      const ks = stringifyPrimitive(key, options.encodeUriComponents) + options.equals!;
      if (value === undefined || value === null) return '';
      if (Array.isArray(value)) {
        const combinedInnerValues: string[] = [];
        for (const innerValue of value) {
          combinedInnerValues.push(ks + stringifyPrimitive(innerValue, options.encodeUriComponents));
        }
        combinedValues.push(combinedInnerValues.join(options.separator));
      } else {
        combinedValues.push(ks + stringifyPrimitive(value, options.encodeUriComponents));
      }
    }

    if (options.includeQuestion) combinedValues[0] = `?${combinedValues[0]}`;

    return combinedValues.join(options.separator);
  } catch (err) {
    if (/object_is_empty/i.test(err.toString())) {
      return `${options.includeQuestion ? '?' : ''}${encodeURIComponent(
        stringifyPrimitive(obj, options.encodeUriComponents)
      )}`;
    }
    if (/input_not_object/i.test(err.toString())) {
      throw '@Favware/Querystring: Your input was not an object. Please supply an object when using Stringify.';
    }

    throw err;
  }
}

/**
 * Parses a querystring back to an object
 *
 * @param qs Querystring to parse
 * @param options Options for the parse, see {@link IParseOptions}
 * @returns A JavaScript object of the parsed parameters
 * @example
 * ```ts
 * parse('?prop=value&prop2=value2')
 * ```
 *
 * @example
 * ```ts
 * parse('?prop=value&prop2=value2', {separator: '&', equals: '='})
 * ```
 *
 * @example
 * ```ts
 * parse('prop=value&&prop2=value2', {separator: '&&', equals: '='})
 * ```
 */
export function parse<R extends Record<PropertyKey, unknown> = Record<PropertyKey, unknown>>(
  qs = '',
  options: ParseOptions = { separator: '&', equals: '=' }
): R {
  try {
    if (typeof qs !== 'string') throw new Error('input_not_string');
    if (qs === '') return {} as R;
    if (/^https?:\/\//.test(qs)) qs = qs.split('?')[1];
    if (!options.separator) options.separator = '&';
    if (!options.equals) options.equals = '=';
    if (qs.startsWith('?')) qs = qs.slice(1);

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const obj = {} as any;
    const regexp = /\+/g;
    const queries: string[] = qs.split(options.separator);
    const maxKeys = 1000;

    let len = queries.length;
    if (maxKeys > 0 && len > maxKeys) len = maxKeys;

    for (const query of queries) {
      const x = query.replace(regexp, '%20');
      const idx = x.indexOf(options.equals);
      let kstr;
      let vstr;

      if (idx >= 0) {
        kstr = x.substr(0, idx);
        vstr = x.substr(idx + 1);
      } else {
        kstr = x;
        vstr = '';
      }

      const k = decodeURIComponent(kstr);
      const v = decodeURIComponent(vstr);

      if (!Reflect.has(obj, k)) {
        obj[k] = v;
      } else if (Array.isArray(obj[k])) {
        obj[k].push(v);
      } else {
        (obj as Record<PropertyKey, unknown>)[k] = [obj[k], v];
      }
    }

    return obj as R;
  } catch (err) {
    if (/input_not_string/i.test(err.toString())) {
      throw '@Favware/Querystring: Your input was not an string. Please supply a string when using Parse.';
    }

    throw err;
  }
}

export default stringify;
