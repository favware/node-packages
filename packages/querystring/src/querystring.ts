import { FunctionSource, ParseOptions, QuerystringError, QuerystringObject, StringifyOptions } from './types';

/**
 * Checks what kind of primitive type the property is and transforms accordingly
 * Used for stringify
 * @param v Input to check for primitive type
 * @private
 */
const stringifyPrimitive = (v: any): string => { // eslint-disable-line @typescript-eslint/no-explicit-any
  if (v === undefined || v === null) return `${v}`;
  switch (typeof v) {
    case 'string':
      return v;
    case 'boolean':
      return v ? 'true' : 'false';
    case 'number':
      return isFinite(v) ? v.toString() : '';
    default:
      return '';
  }
};

/**
 * Gracefully handles errors thrown by other functions
 * @param {string} err The error message to parse
 * @param {FunctionSource} source The source of the error, one of {@link FunctionSource}
 * @private
 */
const handleQuerystringError = (err: string, source: FunctionSource): string | QuerystringError => {
  switch (source) {
    case FunctionSource.Parse:
      return { err };
    case FunctionSource.Stringify:
    default:
      return err;
  }
};

/**
 * Stringifies an object
 * @method
 * @name stringify
 * @param {QuerystringObject} obj Object to stringify
 * @param {StringifyOptions} [options] Options for the stringify, see {@link IStringifyOptions}
 * @returns {string}
 */
export const stringify = (obj: QuerystringObject, options: StringifyOptions = { separator: '&', equals: '=', includeQuestion: false }): string => {
  try {
    if (!obj || Object.keys(obj).length <= 0) throw new Error('object_is_empty');
    if (typeof obj !== 'object') throw new Error('input_not_object');
    if (!options.separator) options.separator = '&';
    if (!options.equals) options.equals = '=';
    if (!options.includeQuestion) options.includeQuestion = false;
    const keys = Object.keys(obj).sort() // eslint-disable-line @typescript-eslint/require-array-sort-compare
      .map(key => {
        const ks = encodeURIComponent(stringifyPrimitive(key)) + options.equals!;
        if (obj[key] === undefined) return ks + encodeURIComponent(stringifyPrimitive(obj[key]));
        if (obj[key] === null) return ks + encodeURIComponent(stringifyPrimitive(obj[key]));
        if (Array.isArray(obj[key])) {
          return obj[key]
            .map((v: string) => ks + encodeURIComponent(stringifyPrimitive(v)))
            .join(options.separator);
        }

        return (
          ks + encodeURIComponent(stringifyPrimitive(obj[key]))
        );
      }).filter(Boolean);

    if (options.includeQuestion) keys[0] = `?${keys[0]}`;

    return keys.join(options.separator);
  } catch (err) {
    if (/object_is_empty/i.test(err.toString())) {
      return `${options.includeQuestion ? '?' : ''}${encodeURIComponent(stringifyPrimitive(obj))}`;
    }
    if (/input_not_object/i.test(err.toString())) {
      return handleQuerystringError(
        '@Favware/Querystring: Your input was not an object. Please supply an object when using Stringify',
        FunctionSource.Stringify
      ) as string;
    }

    return err;
  }
};

/**
 * Parses a querystring back to an object
 * @method
 * @name parse
 * @param {string} qs Querystring to parse
 * @param {ParseOptions} [options] Options for the parse, see {@link IParseOptions}
 * @returns {QuerystringObject}
 */
export const parse = (qs = '', options: ParseOptions = { separator: '&', equals: '=' }): QuerystringObject => {
  try {
    if (typeof qs !== 'string') throw new Error('input_not_string');
    if (qs === '') return {};
    if (/^https?:\/\//.test(qs)) qs = qs.split('?')[1];
    if (!options.separator) options.separator = '&';
    if (!options.equals) options.equals = '=';
    if (qs.startsWith('?')) qs = qs.slice(1);
    const obj: QuerystringObject = {};
    const regexp = /\+/g;
    const queries: string[] = qs.split(options.separator);
    const maxKeys = 1000;

    let len = queries.length;
    if (maxKeys > 0 && len > maxKeys) len = maxKeys;

    for (const query of queries) {
      const x = query.replace(regexp, '%20');
      const idx = x.indexOf(options.equals);
      let kstr;
      let vstr;

      if (idx >= 0) {
        kstr = x.substr(0, idx);
        vstr = x.substr(idx + 1);
      } else {
        kstr = x;
        vstr = '';
      }

      const k = decodeURIComponent(kstr);
      const v = decodeURIComponent(vstr);

      if (!obj.hasOwnProperty(k)) {
        obj[k] = v;
      } else if (Array.isArray(obj[k])) {
        obj[k].push(v);
      } else {
        obj[k] = [ obj[k], v ];
      }
    }

    return obj;
  } catch (err) {
    if (/input_not_string/i.test(err.toString())) {
      return handleQuerystringError(
        '@Favware/Querystring: Your input was not an string. Please supply a string when using Parse',
        FunctionSource.Parse
      ) as QuerystringError;
    }

    return err;
  }
};

export default stringify;