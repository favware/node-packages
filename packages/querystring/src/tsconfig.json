{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "rootDir": "./",
    "outDir": "../dist",
    "composite": true,
    "tsBuildInfoFile": "../dist/.tsbuildinfo",
    "target": "ES5"
  },
  "files": ["index.ts"]
}
