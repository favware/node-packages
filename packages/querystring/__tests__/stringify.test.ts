import { stringify } from '../src';

describe('Standard Behaviour', () => {
  test('should work with empty input', () => {
    const expected = '';
    const actual = stringify({});
    expect(actual).toMatchSnapshot();
    expect(actual).toBe(expected);
  });

  test('should work with empty input and custom options', () => {
    const expected = '?';
    const actual = stringify({}, { separator: '&', equals: '=', includeQuestion: true });
    expect(actual).toMatchSnapshot();
    expect(actual).toBe(expected);
  });

  test('should work without options', () => {
    const expected = 'prop=value&prop2=value2&prop3=true&prop4=6&prop5=false&prop6=&prop7=value1&prop7=value2&prop8=';
    const actual = stringify({
      prop: 'value',
      prop2: 'value2',
      prop3: true,
      prop4: 6,
      prop5: false,
      prop6: Infinity,
      prop7: ['value1', 'value2'],
      prop8: undefined
    });
    expect(actual).toMatchSnapshot();
    expect(actual).toBe(expected);
  });

  test('should exclude props that are undefined or null', () => {
    const expected = 'prop=&prop2=';
    const actual = stringify({
      prop: undefined,
      prop2: null
    });

    expect(actual).toMatchSnapshot();
    expect(actual).toBe(expected);
  });

  test('should work with valid options', () => {
    const expected = '?prop=value&prop2=value2';
    const actual = stringify({ prop: 'value', prop2: 'value2' }, { separator: '&', equals: '=', includeQuestion: true });
    expect(actual).toMatchSnapshot();
    expect(actual).toBe(expected);
  });

  test('confirm default export works', () => {
    const expected = '?prop=value&prop2=value2';
    const actual = stringify({ prop: 'value', prop2: 'value2' }, { separator: '&', equals: '=', includeQuestion: true });
    expect(actual).toMatchSnapshot();
    expect(actual).toBe(expected);
  });

  test('should not encode Uri Components when specified', () => {
    const expected = 'prop=value+_{}":';
    const actual = stringify({ prop: 'value+_{}":' }, { encodeUriComponents: false });
    expect(actual).toMatchSnapshot();
    expect(actual).toBe(expected);
  });

  test('should encode Uri Components when specified', () => {
    const expected = 'prop=value%2B_%7B%7D%22%3A';
    const actual = stringify({ prop: 'value+_{}":' }, { encodeUriComponents: true });
    expect(actual).toMatchSnapshot();
    expect(actual).toBe(expected);
  });
});

describe('Error checking', () => {
  test('should work with wrong options', () => {
    const expected = 'prop%value=prop2%value2';
    const actual = stringify(
      { prop: 'value', prop2: 'value2' },
      // @ts-expect-error This is testing TS-invalid code that is handled in JS
      { separator: '=', equals: '%', includeQeustion: true }
    );
    expect(actual).toMatchSnapshot();
    expect(actual).toBe(expected);
  });

  test('should work with invalid separator and equals options', () => {
    const expected = '?propnullvaluenullprop2nullvalue2';
    const actual = stringify(
      { prop: 'value', prop2: 'value2' },
      // @ts-ignore ts-expect-error doesn't recognize this as a valid usecase yet
      { separator: null, equals: null, includeQuestion: true }
    );
    expect(actual).toMatchSnapshot();
    expect(actual).toBe(expected);
  });

  test('should error with wrong input', () => {
    expect(() => stringify('testing wrong input')).toThrowError(
      '@Favware/Querystring: Your input was not an object. Please supply an object when using Stringify.'
    );
  });
});
