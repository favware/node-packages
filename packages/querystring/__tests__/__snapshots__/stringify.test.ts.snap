// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error checking should work with invalid separator and equals options 1`] = `"?propnullvaluenullprop2nullvalue2"`;

exports[`Error checking should work with wrong options 1`] = `"prop%value=prop2%value2"`;

exports[`Standard Behaviour confirm default export works 1`] = `"?prop=value&prop2=value2"`;

exports[`Standard Behaviour should encode Uri Components when specified 1`] = `"prop=value%2B_%7B%7D%22%3A"`;

exports[`Standard Behaviour should exclude props that are undefined or null 1`] = `"prop=&prop2="`;

exports[`Standard Behaviour should not encode Uri Components when specified 1`] = `"prop=value+_{}\\":"`;

exports[`Standard Behaviour should work with empty input 1`] = `""`;

exports[`Standard Behaviour should work with empty input and custom options 1`] = `"?"`;

exports[`Standard Behaviour should work with valid options 1`] = `"?prop=value&prop2=value2"`;

exports[`Standard Behaviour should work without options 1`] = `"prop=value&prop2=value2&prop3=true&prop4=6&prop5=false&prop6=&prop7=value1&prop7=value2&prop8="`;
